FEC-11577 AdLayout Configuration

GouravSna opened the pull request:
### Description of the Changes

[Works with [Kaltura-Player](https://github.com/kaltura/kaltura-player-android)]

With Ad layout config you can create your own ad break timeline using your vast tags. Ad break can be set as pre, mid and post rolls and each ad break can contain a single vast tag or multiple tags, either as a pod, but also as a Waterfall.

custadian committed:
- Barebone structure for AdLayout

GouravSna committed:
- Handle Waterfalling

GouravSna committed:
- IMAPlugin cleanup

GouravSna committed:
- Code cleanup

GouravSna committed:
- all ads completed masking

GouravSna committed:
- Fixed wrong condition

GouravSna committed:
- Handling multiple AdPods in AdBreak

GouravSna committed:
- Play content player if preroll ad is not configured

giladna committed:
Merge branch 'develop' into FEC-11577-gradlev7

GouravSna committed:
- Fixed issue of Content Playback during next AdPod Loading

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- Fixed glitch of media playback between AdPod playback

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- Fixed possible null exception

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- CuePoints updation

GouravSna committed:
- AdInfo Support for IMAPlugin from AdLayout

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- AdWaterfalling support

GouravSna committed:
- Cuepoints for EVERY based midroll adtype

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -608,25 +617,34 @@ private void cancelAdManagerTimer() {
 
-    private void requestAdsFromIMA(String adTagUrl) {
+    private void requestAdsOnUpdateMedia(String adTagUrl) {
+        log.d("Do requestAdsOnChangeMedia");
```
@GouravSna  please fix log 
requestAdsOnUpdateMedia

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -775,6 +806,52 @@ public void contentCompleted() {
+    @Override
+    public void setAdvertisingConfig(boolean isConfigured, IMAEventsListener imaEventsListener) {
+        isAdvertisingConfigured = isConfigured;
+        this.imaEventsListener = imaEventsListener;
```
@GouravSna  where do you remove this listener when done?

GouravSna replied:
Fixed

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -775,6 +806,52 @@ public void contentCompleted() {
+    @Override
+    public void setCuePoints(List<Long> cuePoints, AdBreakPositionType adBreakPositionType, boolean isUpdatedCuePoint) {
+        if (isAdvertisingConfigured && cuePoints != null && !cuePoints.isEmpty()) {
+            advertisingConfigCuePoints = cuePoints;
```
@GouravSna do you clean it up for change media?

GouravSna replied:
Fixed

GouravSna committed:
- Code refactor for Cuepoints
- Cleanup config on changeMedia
- Support for AdResonse

GouravSna committed:
- Resetted build.gradle file to dev

GouravSna committed:
- Removed line

GouravSna committed:
- Added line

GouravSna committed:
- Removed unused code

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -678,6 +722,7 @@ public void onFinish() {
                     if (getPlayerEngine() != null) {
                         getPlayerEngine().play();
                     }
+                    //TODO: Check if content player should be hidden here
```
what is it?

GouravSna replied:
Removed.

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -608,25 +631,34 @@ private void cancelAdManagerTimer() {
+        }
+    }
+
+    private void requestAdsFromIMA(String adTagResponse, String adTagUrl) {
```
@GouravSna  I think ad tag should be first parameter

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -1144,7 +1269,7 @@ public void onEvent(PlayerEvent.DurationChanged event) {
     private void resetFlagsOnError() {
         isAdError = true;
-        adPlaybackCancelled = true;
+        //adPlaybackCancelled = true; //TODO: Manage state of this flag when waterfalling happens
```
what is this ?

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -1282,7 +1412,11 @@ public void onAdEvent(com.google.ads.interactivemedia.v3.api.AdEvent adEvent) {
+                                    displayAd();
+                                } else {
+                                    getPlayerEngine().play();
+                                }
```
this code can be extracted to method

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

giladna committed:
Merge branch 'develop' into FEC-11577-gradlev7

GouravSna committed:
- Cleanup

giladna committed:
protect NPE

GouravSna committed:
- AutoPlay fixed for AdLayout

GouravSna committed:
- Fix for FEC-11750

GouravSna committed:
- Fire all ads completed if postroll is finished its playback

giladna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -137,6 +141,14 @@
+    private IMAEventsListener imaEventsListener;
+    private boolean isAdvertisingConfigLoading;
+    private List<Long> advertisingConfigCuePoints;
+    private PKAdvertisingAdInfo pkAdvertisingAdInfo = null;
```
@GouravSna  why do we need  = null;

GouravSna committed:
- Spelling correction

GouravSna committed:
- Fixed FEC-11752
- If adlayout is there then marking this boolean true because from controller we are pausing the player before calling playAdNow

GouravSna committed:
- Removed redundant import

GouravSna committed:
- Removed unnecessary values in AdInfo

GouravSna committed:
- Method name refactor

GouravSna commented on imaplugin/src/main/java/com/kaltura/playkit/plugins/ima/IMAPlugin.java:
```diff
@@ -1139,7 +1268,11 @@ public void onEvent(PlayerEvent.DurationChanged event) {
                             IMAPlugin.this.displayContent();
                             if (IMAPlugin.this.getPlayerEngine() != null) {
-                                IMAPlugin.this.getPlayerEngine().play();
+                                if (isAdvertisingConfigLoading()) {
```
Extract to a method

GouravSna replied:
Fixed

GouravSna committed:
- Code refactor
- Extracted method

GouravSna committed:
- Travis build

GouravSna committed:
- Moved to openJdk11 for jitpack

GouravSna committed:
- Travis build

giladna committed:
jitpack build

giladna committed:
jitpack build
